scrunch/mutable_dataset.py

```diff
@@ -0,0 +1,892 @@
+from pycrunch.importing import Importer
+from pycrunch.shoji import wait_progress
+
+from scrunch.expressions import parse_expr, process_expr
+from scrunch.helpers import abs_url, case_expr, subvar_alias
+from scrunch.exceptions import InvalidDatasetTypeError
+from scrunch.datasets import (BaseDataset, _get_connection, _get_dataset,
+                              Variable, CrunchBox)
+from scrunch.subentity import Filter
+from scrunch.variables import (combinations_from_map, combine_categories_expr,
+                               combine_responses_expr, responses_from_map)
+
+_VARIABLE_PAYLOAD_TMPL = {
+    'element': 'shoji:entity',
+    'body': {
+        'name': 'name',
+        'description': 'description',
+        'alias': 'alias'
+    }
+}
+
+_MR_TYPE = 'multiple_response'
+
+
+def get_dataset(dataset, connection=None, editor=False):
+    """
+    A simple wrapper of _get_dataset with streaming=False
+    """
+    shoji_ds, root = _get_dataset(dataset, connection, editor, streaming=False)
+    # make sure the Dataset is of type streaming != "streaming"
+    if shoji_ds['body'].get('streaming') == 'streaming':
+        raise InvalidDatasetTypeError("Dataset %s is of type 'streaming',\
+            use get_streaming_dataset method instead" % dataset)
+    ds = MutableDataset(shoji_ds)
+    if editor is True:
+        ds.change_editor(root.session.email)
+    return ds
+
+
+def create_dataset(name, variables, connection=None):
+    if connection is None:
+        connection = _get_connection()
+        if not connection:
+            raise AttributeError(
+                "Authenticate first with scrunch.connect() or by providing "
+                "config/environment variables")
+
+    shoji_ds = connection.datasets.create({
+        'element': 'shoji:entity',
+        'body': {
+            'name': name,
+            'table': {
+                'element': 'crunch:table',
+                'metadata': variables
+            }
+        }
+    }).refresh()
+    return MutableDataset(shoji_ds)
+
+
+class MutableDataset(BaseDataset):
+    """
+    Class that enclose mutable dataset methods or any
+    method that varies the state of the dataset and/or it's data.
+    """
+
+    def edit(self, **kwargs):
+        """
+        Edit main Dataset attributes
+        """
+        for key in kwargs:
+            if key not in self._MUTABLE_ATTRIBUTES:
+                raise AttributeError("Can't edit attibute %s of variable %s" % (
+                    key, self.name
+                ))
+            if key in ['start_date', 'end_date'] and \
+                    (isinstance(kwargs[key], datetime.date) or
+                     isinstance(kwargs[key], datetime.datetime)
+                     ):
+                kwargs[key] = kwargs[key].isoformat()
+
+        return self.resource.edit(**kwargs)
+
+    def delete(self):
+        """
+        Delete a dataset.
+        """
+        self.resource.delete()
+
+    def stream_rows(self, columns):
```

Comment:
Move this method to the other dataset class. The way things will work is that if a dataset is accepting streaming then it won't accept normal appends, and vice versa.

If a dataset accepts streaming then its schema must be locked (read only dataset). So there will not be streaming support for a mutable dataset.